### ISSUE TITLE
torch.compile(mode="max-autotune") produces different outputs from eager mode

### ISSUE THREAD

**Issue by tinywisdom** [pytorch/pytorch#149472]
### 🐛 Describe the bug

I'm encountering a result mismatch between eager mode and `torch.compile(mode="max-autotune")`.

The outputs differ beyond acceptable tolerances (e.g., `torch.allclose` fails), and this behavior persists in both stable and nightly builds.

### Related Discussion

I initially posted this issue on the PyTorch discussion forum, but have not received a resolution so far.  
Here is the link to the original thread:  
https://discuss.pytorch.org/t/torch-compile-mode-max-autotune-produces-different-inference-result-from-eager-mode-is-this-expected/217873

Since this appears to be a reproducible and version-independent issue, I'm now submitting it here as a formal GitHub issue.

###  Versions

- PyTorch 2.5.1 (original test)
- PyTorch 2.6.0.dev20241112+cu121 (nightly)
- CUDA 12.1
- Platform: Ubuntu 22.04.4 LTS

### Output
=== Detailed comparison ===
- Total number of elements: 3,211,264
- Max absolute error: 0.00128412
- Mean absolute error: 0.000100889
- Max relative error: 23,868.7
- Mean relative error: 0.285904
- Number of elements exceeding tolerance: 98,102
- Percentage of out-of-tolerance elements: 3.05%
- Result of torch.allclose(output_eager, output_compiled, atol=1e-5): False

### Model

Here is my model:
```python
import torch.nn as nn

class BaseConv(nn.Module):
    def __init__(self, in_channels, out_channels, kernel_size, stride, padding, conv_layer):
        super().__init__()
        self.conv = conv_layer(in_channels, out_channels, kernel_size=kernel_size, stride=stride, padding=padding, bias=False)

    def forward(self, x):
        return self.conv(x)

class ActivatedConv(BaseConv):
    def __init__(self, in_channels, out_channels, kernel_size, stride, padding, conv_layer, activation):
        super().__init__(in_channels, out_channels, kernel_size, stride, padding, conv_layer)
        self.activation = activation

    def forward(self, x):
        return self.activation(self.conv(x))

class NormalizedConv(ActivatedConv):
    def __init__(self, in_channels, out_channels, kernel_size, stride, padding, conv_layer, norm, activation):
        super().__init__(in_channels, out_channels, kernel_size, stride, padding, conv_layer, activation)
        self.norm = norm(out_channels)

    def forward(self, x):
        return self.activation(self.norm(self.conv(x)))

class Conv2DBNReLU(NormalizedConv):
    def __init__(self, in_channels, out_channels, kernel_size, stride, padding):
        super().__init__(in_channels, out_channels, kernel_size, stride, padding, nn.Conv2d, nn.BatchNorm2d, nn.ReLU())

class MyModel(nn.Module):
    def __init__(self, in_channels=3, out_channels=64, kernel_size=3, stride=1, padding=1):
        super().__init__()
        self.conv1 = Conv2DBNReLU(in_channels, out_channels, kernel_size, stride, padding)

    def forward(self, x):
        return self.conv1(x)

def my_model_function(in_channels=3, out_channels=64, kernel_size=3, stride=1, padding=1):
    return MyModel(in_channels, out_channels, kernel_size, stride, padding)

if __name__ == "__main__":
    model = my_model_function()
    print(model)
```

### Minimal Script
And this is a minimal script that reproduces the issue:
```python
import torch
import importlib.util
import os

def load_model_from_file(module_path, model_function_name="my_model_function"):
    model_file = os.path.basename(module_path)[:-3]
    spec = importlib.util.spec_from_file_location(model_file, module_path)
    model_module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(model_module)

    model_function = getattr(model_module, model_function_name)
    model = model_function()
    return model



def compare_outputs(a: torch.Tensor, b: torch.Tensor, atol=1e-5, rtol=1e-3):
    print("=== Output difference comparison ===")
    diff = a - b
    abs_diff = diff.abs()
    rel_diff = abs_diff / (a.abs() + 1e-8)
    total_elements = a.numel()
    print(f"- Total elements: {total_elements}")
    print(f"- Max absolute error: {abs_diff.max().item():.8f}")
    print(f"- Mean absolute error: {abs_diff.mean().item():.8f}")
    print(f"- Max relative error: {rel_diff.max().item():.8f}")
    print(f"- Mean relative error: {rel_diff.mean().item():.8f}")
    num_exceed = (~torch.isclose(a, b, atol=atol, rtol=rtol)).sum().item()
    print(f"- Elements exceeding tolerance: {num_exceed}")
    print(f"- Percentage exceeding tolerance: {100.0 * num_exceed / total_elements:.4f}%")
    print(f"- torch.allclose: {torch.allclose(a, b, atol=atol, rtol=rtol)}")



if __name__ == "__main__":
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    input_tensor = torch.rand(1, 3, 224, 224, device=device)

    model_path = "xxx/xxx/xxx/xxx.py"
    model = load_model_from_file(model_path).to(device).eval()

    with torch.no_grad():
        output_eager = model(input_tensor)

    compiled_model = torch.compile(model, mode="max-autotune")
    with torch.no_grad():
        output_compiled = compiled_model(input_tensor)

    compare_outputs(output_eager, output_compiled)
```

### Nightly
```
[pip3] numpy==2.1.2
[pip3] nvidia-cublas-cu12==12.1.3.1
[pip3] nvidia-cuda-cupti-cu12==12.1.105
[pip3] nvidia-cuda-nvrtc-cu12==12.1.105
[pip3] nvidia-cuda-runtime-cu12==12.1.105
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.0.2.54
[pip3] nvidia-curand-cu12==10.3.2.106
[pip3] nvidia-cusolver-cu12==11.4.5.107
[pip3] nvidia-cusparse-cu12==12.1.0.106
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.1.105
[pip3] nvidia-nvtx-cu12==12.1.105
[pip3] pytorch-triton==3.1.0+cf34004b8a
[pip3] torch==2.6.0.dev20241112+cu121
[pip3] torchaudio==2.5.0.dev20241112+cu121
[pip3] torchvision==0.20.0.dev20241112+cu121
[conda] numpy                     2.1.2                    pypi_0    pypi
[conda] nvidia-cublas-cu12        12.1.3.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.1.105                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.0.2.54                pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.2.106               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.4.5.107               pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.1.0.106               pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.6.2                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.21.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.1.105                 pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.1.105                 pypi_0    pypi
[conda] pytorch-triton            3.1.0+cf34004b8a          pypi_0    pypi
[conda] torch                     2.6.0.dev20241112+cu121          pypi_0    pypi
[conda] torchaudio                2.5.0.dev20241112+cu121          pypi_0    pypi
[conda] torchvision               0.20.0.dev20241112+cu121          pypi_0    pypi
```

### Original
```
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.5.3.2
[pip3] nvidia-cuda-cupti-cu12==12.5.82
[pip3] nvidia-cuda-nvrtc-cu12==12.5.82
[pip3] nvidia-cuda-runtime-cu12==12.5.82
[pip3] nvidia-cudnn-cu12==9.3.0.75
[pip3] nvidia-cufft-cu12==11.2.3.61
[pip3] nvidia-curand-cu12==10.3.6.82
[pip3] nvidia-cusolver-cu12==11.6.3.83
[pip3] nvidia-cusparse-cu12==12.5.1.3
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.5.82
[pip3] optree==0.13.1
[pip3] torch==2.5.1
[pip3] torchaudio==2.5.1
[pip3] torchdata==0.10.0
[pip3] torchvision==0.20.1
[pip3] triton==3.1.0
[conda] blas                      1.0                         mkl    defaults
[conda] cuda-cudart               12.1.105                      0    nvidia
[conda] cuda-cupti                12.1.105                      0    nvidia
[conda] cuda-libraries            12.1.0                        0    nvidia
[conda] cuda-nvrtc                12.1.105                      0    nvidia
[conda] cuda-nvtx                 12.1.105                      0    nvidia
[conda] cuda-opencl               12.6.77                       0    nvidia
[conda] cuda-runtime              12.1.0                        0    nvidia
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libblas                   3.9.0            16_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            16_linux64_mkl    conda-forge
[conda] libcublas                 12.1.0.26                     0    nvidia
[conda] libcufft                  11.0.2.4                      0    nvidia
[conda] libcurand                 10.3.7.77                     0    nvidia
[conda] libcusolver               11.4.4.55                     0    nvidia
[conda] libcusparse               12.0.2.55                     0    nvidia
[conda] liblapack                 3.9.0            16_linux64_mkl    conda-forge
[conda] libnvjitlink              12.1.105                      0    nvidia
[conda] mkl                       2022.1.0           hc2b9512_224    defaults
[conda] numpy                     1.26.4          py310hb13e2d6_0    conda-forge
[conda] nvidia-cublas-cu12        12.5.3.2                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.5.82                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.5.82                  pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.5.82                  pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.3.0.75                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.2.3.61                pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.6.82                pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.6.3.83                pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.5.1.3                 pypi_0    pypi
[conda] nvidia-nccl-cu12          2.21.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.5.82                  pypi_0    pypi
[conda] optree                    0.13.1                   pypi_0    pypi
[conda] pytorch                   2.5.1           py3.10_cuda12.1_cudnn9.1.0_0    pytorch
[conda] pytorch-cuda              12.1                 ha16c6d3_6    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                2.5.1               py310_cu121    pytorch
[conda] torchdata                 0.10.0                   pypi_0    pypi
[conda] torchtriton               3.1.0                     py310    pytorch
[conda] torchvision               0.20.1              py310_cu121    pytorch

```

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @aakhundov

**Comment by malfet** [comment]
`torch.allclose` default tolerances are too narrow for any reduction/normalization op, so those discrepancies might be acceptable

**Comment by tinywisdom** [comment]
> `torch.allclose` default tolerances are too narrow for any reduction/normalization op, so those discrepancies might be acceptable

Thank you for the clarification — that makes sense!

Just to better understand the boundary here:  
***Is there a recommended threshold or guideline for what level of discrepancy is considered unacceptable?***

For example, is it more appropriate to use absolute error, relative error, or some combination (e.g., `torch.allclose(a, b, atol=..., rtol=...)`)?

In my case, I'm seeing relative errors up to ~20,000 and ~3% of the elements exceeding `torch.allclose` with the default tolerances.  
Would a discrepancy of this magnitude still be considered acceptable for typical reduction/normalization ops?

Any insights on how to choose reasonable `atol` / `rtol` values for comparisons in compiled vs. eager mode would be very helpful!

**Comment by IvanKobzarev** [comment]
@tinywisdom 
In your case you are getting atol=2e-3 but the relative error is very high, I would guess this is because of small numbers in the output.

In https://github.com/pytorch/pytorch/blob/main/test/inductor/test_max_autotune.py you can see that using atol=1e-2 is pretty common.


I do not know any deterministic algorithm, how to calculate atol/rtol  for each case. In our tests they are all different per example. 

As for recommendations - to use float64 calculations as a reference value.
Could you try float64 model and downsize the model and to check what will be the `rtol` in that case?

**Comment by tinywisdom** [comment]
> [@tinywisdom](https://github.com/tinywisdom) In your case you are getting atol=2e-3 but the relative error is very high, I would guess this is because of small numbers in the output.
> 
> In https://github.com/pytorch/pytorch/blob/main/test/inductor/test_max_autotune.py you can see that using atol=1e-2 is pretty common.
> 
> I do not know any deterministic algorithm, how to calculate atol/rtol for each case. In our tests they are all different per example.
> 
> As for recommendations - to use float64 calculations as a reference value. Could you try float64 model and downsize the model and to check what will be the `rtol` in that case?

Thank you very much for the detailed explanation — that was really helpful.

I'll definitely keep your suggestions in mind when analyzing future discrepancies, especially regarding tolerances and using float64 as a reference.

Much appreciated!

Thanks again — I'll experiment with float64 and downscaled inputs to better understand how the tolerances behave.
